crypto/dsa/dsa_ossl.c

```diff
@@ -201,9 +201,9 @@ static int dsa_sign_setup(DSA *dsa, BN_CTX *ctx_in,
 
     /* Preallocate space */
     q_bits = BN_num_bits(dsa->q);
-    if (!BN_set_bit(k, q_bits)
-        || !BN_set_bit(l, q_bits)
-        || !BN_set_bit(m, q_bits))
+    q_words = bn_get_top(dsa->q);
+    if (!bn_wexpand(k, q_words + 2)
+        || !bn_wexpand(l, q_words + 2))
```

Comment:
wouldn't q_words + 1 be sufficient to store k, k+q  and k+2q ?

Reply:
No, because of this code in BN_uadd:

https://github.com/openssl/openssl/blob/8817215d5c52a76f2b184b624bde4df8556dee6d/crypto/bn/bn_add.c#L85-L97

Take the case where q == 160 bits == 5 words. After:

````
    BN_add(l, k, dsa->q)
````
`l->top` will either be 5 or 6 depending on the value of `k`. After:

````
    BN_add(k, l, dsa->q))
````
If `l->top == 5` then `BN_uadd()` will attempt to allocate 5+1 words. If `l->top == 6` then `BN_uadd()` will attempt to allocate 6+1 words (`== q_words + 2`).

Reply:
Ok, thanks.

Reply:
Thanks, Matt.  Saved me trying to explain.